### PR TITLE
Move tsconfigs logic to Node

### PR DESCRIPTION
**Pull request by victor-diez-sonarsource**

### CODE REVIEW

File: src/helpers/lru.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+export class LRU<T> {
+  private readonly max: number;
+  private readonly cache: T[];
+  constructor(max = 2) {
+    this.max = max;
+    this.cache = [];
+  }
+
+  get() {
+    return this.cache;
```
Comment (ilia-kebets-sonarsource) [inline src/helpers/lru.ts:29]: aren't reads supposed to count as usage?

File: src/helpers/tsconfigs.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ */
+import fs from 'fs';
+import path from 'path';
+import { getContext } from './context';
+import { readFileSync, toUnixPath } from './files';
+import * as console from 'console';
+import { defaultCache } from 'services/program';
+
+export type TSConfigs = Map<string, TSConfig>;
+export const projectTSConfigs: TSConfigs = new Map<string, TSConfig>();
```
Comment (ilia-kebets-sonarsource) [inline src/helpers/tsconfigs.ts:28]: somehow specify that it's a singleton

File: tests/helpers/tsconfigs.test.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+import { projectTSConfigs, setContext, toUnixPath, tsConfigLookup, updateTsConfigs } from 'helpers';
+import path from 'path';
+import console from 'console';
+
+describe('TSConfigs', () => {
```
Comment (ilia-kebets-sonarsource) [inline tests/helpers/tsconfigs.test.ts:25]: specify the function you're testing like `tsConfigLookup()` or `updateTsConfigs()`

File: src/helpers/lru.ts (hunk 1 above)
Comment (victor-diez-sonarsource) [reply src/helpers/lru.ts:29]: The GC only collects WeakRefs. LRU is used to keep strong refs and avoid GC. I can explain this on a call, but no worries I'll further comment on the code to clarify it better

File: src/helpers/tsconfigs.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+      projectTSConfigs.set(filename, {
+        filename,
+        contents,
+      });
+    }
+  }
+}
+
+export function updateTsConfigs(tsconfigs: string[], force = false) {
+  if (!getContext().sonarlint && !force) {
```
[the remark this answers is not in the document]
Comment (victor-diez-sonarsource) [reply src/helpers/tsconfigs.ts:61]: yes, it's just that negating the condition and moving the whole code inside increments the cognitive complexity of the function. :(

File: src/helpers/cache.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ */
+import { LRU } from './lru';
+import ts from 'typescript';
+import { TSConfig } from './tsconfigs';
+
+export type ProgramResult = {
+  tsConfig: TSConfig;
+  files: string[];
+  projectReferences: string[];
+  missingTsConfig: boolean;
```
Comment (ilia-kebets-sonarsource) [inline src/helpers/cache.ts:28]:
```suggestion
  isMissingTsConfig: boolean;
```

File: src/helpers/cache.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  program: WeakRef<ts.Program>;
+  fallbackProgram?: boolean;
```
Comment (ilia-kebets-sonarsource) [inline src/helpers/cache.ts:30]:
```suggestion
  isFallbackProgram?: boolean;
```

File: src/helpers/tsconfigs.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+export type TSConfigs = Map<string, TSConfig>;
+export const projectTSConfigs: TSConfigs = new Map<string, TSConfig>();
+export interface TSConfig {
+  filename: string;
+  contents: string;
+  fallbackTSConfig?: boolean;
+}
+
+export function tsConfigLookup(dir?: string) {
```
Comment (ilia-kebets-sonarsource) [inline src/helpers/tsconfigs.ts:35]: WDYT about changing this from DFS to BFS, and look for `index.(js|ts)` and `main.(js|ts)` files before iterating on all as we discussed the other day?

File: src/helpers/tsconfigs.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ilia-kebets-sonarsource) [reply src/helpers/tsconfigs.ts:61]: nevermind indeed.

File: src/helpers/tsconfigs.ts (hunk 7 above)
Comment (victor-diez-sonarsource) [reply src/helpers/tsconfigs.ts:35]: I would prefer to leave the ordering aside of the lookup. I think ordering algorithm should be implemented [here](https://github.com/SonarSource/SonarJS/blob/e8c595efe085a262b18068674589f64f083262c6/src/services/program/program.ts#L50)

File: src/parsing/jsts/builders/build.ts [hunk 8]
```diff
@@ -35,22 +35,79 @@ import { Language } from 'parsing/jsts';
...
+     */
+    throw moduleError;
   }
 }
 
-function shouldTryTypeScriptParser() {
-  const context = getContext();
-  return context ? context.shouldUseTypeScriptParserForJS : true;
+function shouldUseWatchProgram(file: string): boolean {
+  return getContext()?.sonarlint || file.toLowerCase().endsWith('.vue');
```
Comment (ilia-kebets-sonarsource) [inline src/parsing/jsts/builders/build.ts:108]: you can use `isVueFile` computed before as function param

File: src/helpers/cache.ts (hunk 5 above)
Comment (victor-diez-sonarsource) [reply src/helpers/cache.ts:28]: I would leave this by now as it's used currently from the Java part.

**Review by saberduck — APPROVED** [review]

**Pull request merged**